demos/mqtt/mqtt_demo_connection_sharing/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,308 @@
+                        const uint16_t topicFilterLength,
+                        bool requiresExactMatch )
+{
+    bool status = false;
+
+    /* Check for an exact match. */
+    if( topicNameLength == topicFilterLength )
+    {
+        status = ( strncmp( pTopicName, pTopicFilter, topicNameLength ) == 0 );
+    }
+
+    /* If  an exact match is required, return the result of the comparison above.
+     * Otherwise, attempt to match with MQTT wildcards in topic filters. */
+    if( requiresExactMatch == false )
+    {
+        status = topicFilterMatch( pTopicName, topicNameLength, pTopicFilter, topicFilterLength );
+    }
+
+    return status;
+}
+
+/* Iterate in the record list till the first record (from the starting index passed) has a topic match. */
+bool findRecordForTopic( const char * pTopicName,
+                         size_t topicNameLength,
+                         size_t * pCursor )
+{
+    bool matchFound = false;
+
+    do
+    {
+        matchFound = topicMatch( pTopicName,
+                                 topicNameLength,
+                                 callbackRecordList[ *pCursor ].pTopicFilter,
+                                 callbackRecordList[ *pCursor ].topicFilterLength,
+                                 callbackRecordList[ *pCursor ].requiresExactMatch );
+        /* code */
+    } while( ++( *pCursor ) < recordListSize && ( matchFound == false ) );
+
+    return matchFound;
+}
+
+
+/* Design -
+ *   * Common handler can contain logic of processing acks and retries.
+ *   * Only forward PUBLISH message to message dispatcher/handler as individual callbacks
+ *     only want to process incoming PUBLISH message.
+ */
+void SubscriptionManager_DispatchHandler( uint16_t packetIdentifier,
+                                          MQTTPublishInfo_t * pPublishInfo )
+{
+    assert( pPublishInfo != NULL );
+    assert( packetIdentifier != MQTT_PACKET_ID_INVALID );
+
+    size_t recordListCursor = 0u;
+
+    while( findRecordForTopic( pPublishInfo->pTopicName,
+                               pPublishInfo->topicNameLength,
+                               &recordListCursor ) != false )
+    {
+        callbackRecordList[ recordListCursor ].callback( packetIdentifier, pPublishInfo );
+    }
+}
+
+bool SubscriptionManager_RegisterCallback( const char * pTopicFilter,
+                                           size_t topicFileterLength,
+                                           SubscriptionManager_Callback_t callback )
+{
+    bool recordAdded = false;
+    size_t availableIndex = 0u;
+
+    /* Search for an available spot in the list to store the record */
+    while( ( callbackRecordList[ availableIndex ].pTopicFilter != NULL ) &&
+           ( availableIndex < recordListSize ) )
+    {
+        availableIndex++;
+    }
+
+    if( availableIndex == recordListSize )
+    {
+        /* Record list is full. */
+    }
+    else
+    {
+        /* Should the topic string be copied? */
```

Comment:
To accommodate varying use cases yes, so the `pTopicFilter` member of the `SubscriptionManager_Record_t` would need to be an array. However, since it's just a demo and you control everything, it's up to you; if you only use constant strings then I don't think it's a major issue as long as there's a comment calling that out.

Reply:
Yeah, I think for the single-threaded demo purpose, I will keep it simple and avoid the memory allocation problem (currently, the record list is statically allocated for a fixed size, but with local copy of the topic filter, it would require a max limit on the topic filter size, and thereby, increase the record list structure size)